peps/pep-0755.rst

```diff
@@ -0,0 +1,907 @@
+
+-  ``Self`` (valid only in some contexts)
+-  ``TypeGuard[...]`` (valid only in some contexts)
+-  ``TypeIs[...]`` (valid only in some contexts)
+
+``Self`` can be used as a function parameter type, a return type, or a
+variable type, but only within a class definition:
+
+::
+
+   class ReturnsSelf:
+       def __call__(self: Self) -> Self:  # OK
+           return self
+
+   class HoldsSelf:
+       me: Self  # OK
+
+   def returns_self(me: Self) -> Self: ...  # ERROR: Self meaningless outside class
+
+``TypeGuard[]`` and ``TypeIs[]`` can only be used as a return type of a
+function. They cannot be used as a variable type or a parameter type:
+
+::
+
+   def is_positive_int(value: object) -> TypeGuard[int]: ...  # OK
+
+   def nonsense(value: TypeGuard[int]): ...  # ERROR: TypeGuard[] not meaningful here
+
+   exotic_bool: TypeGuard[int]  # ERROR: TypeGuard[] not meaningful here
+
+Non-universal type expressions may be treated as equivalent to ``Any``
+for inference purposes:
+
+::
+
+   typx: TypeExpr = TypeIs[str]
+   assert_type(typx, TypeExpr[Any])  # acceptable
+
+Non-universal type expressions may also be treated as equivalent to
+``Any`` when 
+:ref:`constraining type variables <interactions_with_type_variables>`.
+
+
+Static vs. Runtime Representations of TypeExprs
+'''''''''''''''''''''''''''''''''''''''''''''''
+
+A TypeExpr value appearing statically in a source file may be normalized
+to a different representation at runtime:
+
+-  For example string-based forward references are normalized at runtime
+   to be ``ForwardRef`` instances: [#forward_ref_normalization]_
+
+   ::
+
+      >>> IntTree = list[typing.Union[int, 'IntTree']]
+      >>> IntTree
+      list[typing.Union[int, ForwardRef('IntTree')]]
+
+-  Special forms themselves often have a runtime representation which is
+   different than how they appear in a source file:
+
+   ::
+
+      >>> list[int]
+      list[int]
+      >>> type(list[int])
+      <class 'types.GenericAlias'>
+      >>> types.GenericAlias(list, int) == list[int]
+      True
+
+The runtime representations of TypeExprs are considered implementation
+details that may change over time and therefore static type checkers are
+not required to recognize them:
+
+::
+
+   INT_TREE: TypeExpr = ForwardRef('IntTree')  # ERROR: Runtime-only form
+   LIST_OF_INT: TypeExpr = types.GenericAlias(list, int)  # ERROR: Runtime-only form
+
+However runtime type checkers must still be prepared to recognize and
+manipulate such runtime-only representations.
+
+Runtime type checkers that wish to assign a runtime-only representation
+of a type expression to a TypeExpr[] variable must use ``cast()`` to
+avoid errors from static type checkers:
+
+::
+
+   INT_TREE = cast(TypeExpr, ForwardRef('IntTree'))  # OK
+   LIST_OF_INT = cast(TypeExpr, types.GenericAlias(list, int))  # OK
+
+
+Subtyping
+---------
+
+Whether a TypeExpr value can be assigned from one variable to another is
```

Comment:
Consistently apply code formatting for things like this throughout the document:
```suggestion
Whether a ``TypeExpr`` value can be assigned from one variable to another is
```